Support CORS request

CzBiX opened the pull request:
This patch allows authentication information to be passed via the http header, which is the session id.

It is mainly for allow CORS request, so [my WebUI](https://github.com/CzBiX/qb-web/wiki/How-to-use) can be used by the user online without locally installation. But cookies have security restrictions that cannot be bypassed in CORS(See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite). There may also be new security restrictions on cookies in the future, so I tend to change the http header. This mechanism may also be used for the next generation API architecture.

This patch also can simplifying the writing of scripts without handling of cookies. Since not all HTTP libraries support handle cookies(such as `urllib` in python).

This PR can be tested with:
```sh
# get token
curl -X POST -F 'username=admin' -F 'password=adminadmin' -F 'token_mode=1' http://localhost:8080/api/v2/auth/login
# check the token works
curl -H 'Authorization: Bearer foobar' http://localhost:8080/api/v2/app/versioin
```

glassez reviewed (changes requested):
There are multiple Coding Style violations that should be fixed.

Piccirello commented on src/webui/webapplication.cpp:
```diff
@@ -610,28 +615,35 @@ void WebApplication::sessionStart()
-    setHeader({Http::HEADER_SET_COOKIE, QString::fromLatin1(cookieRawForm)});
+    const auto sessionId = m_currentSession->id().toUtf8();
+    if (m_request.headers.contains(HEADER_API_TOKEN)) {
+        setHeader({HEADER_API_TOKEN, QString::fromLatin1(sessionId)});
```
Why do we need to echo the token back in the response?

Piccirello replied:
Based on the PR description, this is for retrieving the initial token value. It might be worthwhile to have an explicit endpoint for this, rather than requiring the user to specify a blank auth header in the request.

glassez replied:
>Based on the PR description, this is for retrieving the initial token value.

IIRC, this is for asking the server to use tokens instead of cookies.

CzBiX replied:
Added new `auth/auth` endpoint for it.

Piccirello commented on src/webui/webapplication.cpp:
```diff
@@ -65,6 +65,7 @@
 const int MAX_ALLOWED_FILESIZE = 10 * 1024 * 1024;
 const auto C_SID = QByteArrayLiteral("SID"); // name of session id cookie
 
+const QString HEADER_API_TOKEN = u"x-api-token"_qs; // name of api token header
```
I think the standard `Authorization` header would be more appropriate here, with a scheme of `Bearer`.

CzBiX replied:
Switched to `Bearer` scheme.

CzBiX commented:
> There are multiple Coding Style violations that should be fixed. See https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md

Sorry I missed that. I tried use uncrustify to format code, found that this file has many lines that don't match the code style, should I submit only the lines I changed or submit the whole file?

glassez commented:
> should I submit only the lines I changed or submit the whole file?

Sorry, I'm not clear what you meant. If you just fix the Coding Style, then you can amend existing commit and force push it.

CzBiX commented:
> > should I submit only the lines I changed or submit the whole file?
> 
> Sorry, I'm not clear what you meant. If you just fix the Coding Style, then you can amend existing commit and force push it.

Run uncrustify on this file will get a diff about 56 insertions and 106 deletions, but most of the lines are not touched by me.

I already amended new commits that fixes the coding style, but only the parts that I have changed. Hope this is what you expected.

CzBiX commented:
More tests found that the `OPTIONS` method needs to be supported to support CORS. mark this as draft for now.

glassez commented on src/webui/api/authcontroller.h:
```diff
@@ -47,12 +47,14 @@ class AuthController : public APIController
 
 private slots:
     void loginAction();
+    void authAction();
```
IMO, having two actions/endpoints for the same thing (login) isn't a good idea.

CzBiX replied:
Changed to `token_mode` param in login action.

glassez commented:
> Run uncrustify on this file will get a diff about 56 insertions and 106 deletions, but most of the lines are not touched by me.

Doesn't seem that uncrustify config is well maintained.


> I already amended new commits that fixes the coding style, but only the parts that I have changed. Hope this is what you expected.

It's still not fixed. At least the braces are placed incorrectly.

CzBiX commented:
> Doesn't seem that uncrustify config is well maintained.

OK, I format code manually, should be fixed now.

Piccirello commented:
My understanding of this new functionality is that it'll be used by APIs. That's outside of the browser context, so why is CORS support needed?

glassez commented:
> My understanding of this new functionality is that it'll be used by APIs. That's outside of the browser context, so why is CORS support needed?

Substituting the name of the feature (Support CORS request) in your comment, we get some nonsense... Can you clarify yourself?

CzBiX commented:
@glassez @Piccirello any updates?

glassez commented:
> @glassez @Piccirello any updates?

I'm neutral about the feature itself. So I asked others about general approval of it. All we have since then is comment if @Piccirello, which is incomprehensible to me, and which he did not bother to explain in response to my request. @CzBiX, if it's clear to you what he's talking about, you could continue this discussion.

@Chocobo1, what's your opinion about this feature generally?

Chocobo1 commented:
>@Chocobo1, what's your opinion about this feature generally?

Not sure, it sounds like altering the authentication process... which is something I'm not familiar with. I'll need more info before making a decision on this.

CzBiX commented:
Google will ban third-party cookies eventually, and the cookies really aren't good way for API authentication.

https://developer.chrome.com/en/docs/privacy-sandbox/third-party-cookie-phase-out/
